User Defined Functions tutorial (#55)
streamshub/flink-sql-examples: docs/user-defined-functions/index.md

+++
title = 'Simple User Defined Functions'
+++

> Note: This tutorial is mainly focused on creating a simple [Flink SQL](https://nightlies.apache.org/flink/flink-docs-release-1.20/docs/dev/table/overview/) [User Defined Function (UDF)](https://nightlies.apache.org/flink/flink-docs-release-2.0/docs/dev/table/functions/udfs/). For detailed information on working with [Flink ETL Jobs](https://nightlies.apache.org/flink/flink-docs-release-2.0/docs/learn-flink/etl/) and [Session Clusters](https://nightlies.apache.org/flink/flink-kubernetes-operator-docs-main/docs/custom-resource/overview/#session-cluster-deployments), look at the [Interactive ETL example](../interactive-etl/index.md).

> [Flink SQL](https://nightlies.apache.org/flink/flink-docs-release-1.20/docs/dev/table/overview/) is a powerful tool for data exploration, manipulation and inter-connection.

Author: Are you talking about pointing to something different than `nightlies.apache.org`? I can't find any other domain that hosts the docs.

> [Flink SQL](https://nightlies.apache.org/flink/flink-docs-release-1.20/docs/dev/table/overview/) is a powerful tool for data exploration, manipulation and inter-connection.

Author: https://github.com/streamshub/flink-sql-examples/pull/55/commits/52ff05d53bd8b4017754a89c22241c08842513a8

## Scenario

### Source Data Table

The data generator application creates a topic (`flink.international.sales.records`) containing international sales records.
The schema for this topic can be seen in the `data-generator/src/main/resources/internationalSales.avsc`:

```avroschema
{
  "namespace": "com.github.streamshub.kafka.data.generator.schema",
  "type": "record",
  "name": "InternationalSales",
  "fields": [
    {"name": "user_id", "type": "string"},
    {"name": "product_id",  "type": "string"},
    {"name": "invoice_id",  "type": "string"},
    {"name": "quantity",  "type": "string"},
    {"name": "unit_cost",  "type": "string"}
  ]
}
```

However, it looks like the person who owns this schema repeated the same mistake he did for the Sales schema we looked at in the [Interactive ETL example](../interactive-etl/index.md), and decided to once again include the currency symbol at the start of the `unit_cost` field! (at least he's consistent...).

> ```shell

Author: ```suggestion
*(Assuming you have the data generator up and running as per the instructions in the [Setup](#setup) section, you can verify this by running the following command):*

```shell
```

$ kubectl exec -it my-cluster-dual-role-0 -n flink -- /bin/bash \
    ./bin/kafka-console-consumer.sh --bootstrap-server localhost:9092 \
        --topic flink.international.sales.records

user-82130&53972644729678620433
                               ₹192
user-1619&74772726277194883031
                              ฿638
...
```

Instead of having to deal with [Unicode currency symbols](https://www.unicode.org/charts/nameslist/n_20A0.html) in our Flink SQL queries, we can create a simple UDF to strip the currency symbol from the `unit_cost` field and add an equivalent [ISO 4217](https://www.iso.org/iso-4217-currency-codes.html) currency code to the end of the field instead e.g. "€192" will become "192 EUR".

> We will basically be replacing a cumbersome to maintain query e.g.:

Author: https://github.com/streamshub/flink-sql-examples/pull/55/commits/1ce56b3abf42cc883fbc93782894ff2ec86e60fb

        ELSE CONCAT(SUBSTRING(unit_cost, 2), ' ???')
    END AS iso_unit_cost,
    unit_cost,
    quantity
FROM InternationalSalesRecordTable;
```

with this:

```sql
SELECT
    invoice_id,
    currency_convert(unit_cost) AS iso_unit_cost,
    unit_cost,
    quantity
FROM InternationalSalesRecordTable;
```

## Creating the User Defined Function

> Note: You can find the full completed code for the UDF in the `tutorials/currency-converter` directory.

### Creating the Maven project

First, we will create a Maven project in our home directory, and `cd` into it:

```shell
cd ~

mvn archetype:generate \
    -DgroupId=com.github.example \
    -DartifactId=currency-converter \
    -DarchetypeArtifactId=maven-archetype-quickstart \
    -DarchetypeVersion=1.5 \
    -DinteractiveMode=false

cd ~/currency-converter
```

> Note: Flink provides a [Maven Archetype and quickstart script](https://nightlies.apache.org/flink/flink-docs-release-2.0/docs/dev/configuration/overview/#getting-started) for getting started. However, it includes a lot of dependencies and boilerplate we don't need for this tutorial, so we will start with a minimal Maven project instead.

> We can remove the provided tests and their dependencies, as we don't need them for this tutorial:

Author: Should I go through the process of writing the tests in the tutorial?

> We can remove the provided tests and their dependencies, as we don't need them for this tutorial:

Reviewer: If testing the UDF with Flink is a bit more complicated than importing a library and setting up a class it might be worth describing that.

```shell
rm -r src/test

sed -i -e '/<dependencyManagement>/,/<\/dependencyManagement>/d' pom.xml

sed -i -e '/<dependencies>/,/<\/dependencies>/d' pom.xml
```

Your `pom.xml` should now look something like this:

```xml
<?xml version="1.0" encoding="UTF-8"?>
<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
  xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
  <modelVersion>4.0.0</modelVersion>

  <groupId>com.github.example</groupId>
  <artifactId>currency-converter</artifactId>
  <version>1.0-SNAPSHOT</version>

  <name>currency-converter</name>

  <properties>
    <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
    <maven.compiler.release>17</maven.compiler.release>
  </properties>

  <build>
    <pluginManagement><!-- lock down plugins versions to avoid using Maven defaults (may be moved to parent pom) -->
      <plugins>
        <!-- clean lifecycle, see https://maven.apache.org/ref/current/maven-core/lifecycles.html#clean_Lifecycle -->
        <plugin>
          <artifactId>maven-clean-plugin</artifactId>
          <version>3.4.0</version>

>         </plugin>

Reviewer: Actually I am not sure showing the pom is very useful, you have an example in the repo.

public class CurrencyConverter extends ScalarFunction {}
```

This function doesn't do anything yet. For that, we need it to declare a public `eval` method.

Since we'll only be passing it one argument (the `unit_cost` field), we can declare that the method takes in a single `String` argument and also returns a `String`:

```java
package com.github.example;

import org.apache.flink.table.functions.ScalarFunction;

public class CurrencyConverter extends ScalarFunction {
   // (You can name the parameter whatever you like)
   // e.g. currencyAmount = "€100"
   public String eval(String currencyAmount) {
       // logic will go here
   }
}
```

Flink's [Automatic Type Inference](https://nightlies.apache.org/flink/flink-docs-release-2.0/docs/dev/table/functions/udfs/#automatic-type-inference) will use reflection to derive SQL data types for the argument and result of our UDF. If you want to override this behavior, you can [explicitly specify the types]((https://nightlies.apache.org/flink/flink-docs-release-2.0/docs/dev/table/functions/udfs/#automatic-type-inference)), but in this case we will keep it simple and let Flink decide for us.

> If we look at the [Data Generator](https://github.com/streamshub/flink-sql-examples/blob/main/data-generator/src/main/java/com/github/streamshub/kafka/data/generator/examples/InternationalSalesData.java) in the StreamsHub [Flink SQL Examples](https://github.com/streamshub/flink-sql-examples) repository, we can see the possible currency symbols that can appear in the `unit_cost` field:

Author: https://github.com/streamshub/flink-sql-examples/pull/55/commits/cf0e74bc0cc2dd501c7437a561147908fbf4ed66

```java
public class InternationalSalesData implements Data {
    private static final char[] CURRENCY_SYMBOLS = {'€', '₹', '₺', '฿', '₴', '₮'};
    // ... other fields and methods
}
```

In our UDF, we can create a `Map` of these symbols to their corresponding [ISO 4217](https://www.iso.org/iso-4217-currency-codes.html) currency codes. We will use these when converting the `unit_cost` field into our desired format.

```java
package com.github.example;

import java.util.Map;

import org.apache.flink.table.functions.ScalarFunction;

public class CurrencyConverter extends ScalarFunction {
   // https://www.unicode.org/charts/nameslist/n_20A0.html
   // https://www.iso.org/iso-4217-currency-codes.html

>    private static final Map<Character, String> CURRENCY_SYMBOL_ISO_MAP = Map.of(

Author: https://github.com/streamshub/flink-sql-examples/pull/55/commits/d43dc83e69bc4bfb1eee6e4c8e110997c3e0670c

   // https://www.unicode.org/charts/nameslist/n_20A0.html
   // https://www.iso.org/iso-4217-currency-codes.html
   private static final Map<Character, String> CURRENCY_SYMBOL_ISO_MAP = Map.of(
      '€', "EUR",
      '₹', "INR",
      '₺', "TRY",
      '฿', "THB",
      '₴', "UAH",
      '₮', "MNT"
   );

   // Value of passed field (e.g. "unit_cost") is passed in e.g. "€100"
   public String eval(String currencyAmount) {
      // 1. Get the first character of the string, which is the currency symbol (e.g. '€').
      char currencySymbol = currencyAmount.charAt(0);

      // 2. Get the rest of the string, which is the amount (e.g. "100").
      String amount = currencyAmount.substring(1);

      // 3. Look up the currency symbol in our Map to get the corresponding currency code (e.g. '€' => "EUR").
      String currencyIsoCode = CURRENCY_SYMBOL_ISO_MAP.get(currencySymbol);

      // 4. If the currency symbol is not found in the Map, we can return "???" as the currency code.
      if (currencyIsoCode == null) {
         currencyIsoCode = "???";
      }

      // 5. Concatenate the currency code to the amount, and return the result (e.g. "100 EUR").

>       return amount + " " + currencyIsoCode;

Author: https://github.com/streamshub/flink-sql-examples/pull/55/commits/4a3e190261f593656b9ba4e8a879003f1824e7cc

   }
}
```

### Building the JAR

After implementing the logic, we can build our JAR:

```shell
mvn clean package
```

> There should be no compilation errors, and there should be a JAR in the `target` directory.

Author: https://github.com/streamshub/flink-sql-examples/pull/55/commits/adb7c8014587037536f56e127d65b0c299b2301e

    ```shell
    kubectl exec -it my-cluster-dual-role-0 -n flink -- /bin/bash \
    ./bin/kafka-console-consumer.sh --bootstrap-server localhost:9092 --topic flink.international.sales.records
    ```

4. Deploy a [Flink session cluster](https://nightlies.apache.org/flink/flink-kubernetes-operator-docs-main/docs/custom-resource/overview/#session-cluster-deployments):

    ```shell
    kubectl -n flink apply -f user-defined-functions/flink-session-udf.yaml
    ```

### Running the UDF

Since we want to try the UDF in different scenarios, we will create a container containing the Flink SQL CLI to run our queries (see the [Interactive ETL example](../interactive-etl/index.md) for details on the CLI).

First, we need to port forward the Flink Job Manager pod so the Flink SQL CLI can access it:

```shell
kubectl -n flink port-forward <job-manager-pod> 8081:8081
```

The job manager pod will have the name format `session-cluster-udf-<alphanumeric>`, your `kubectl` should tab-complete the name. 
If it doesn't then you can find the job manager name by running `kubectl -n flink get pods`.

Next, we will create a container with our JAR mounted into it:

```shell

> podman run -it --rm --net=host \

Author: https://github.com/streamshub/flink-sql-examples/pull/55/commits/fa5b023e6cb6c5420e62b4f62a837a8b38c342f8

https://github.com/streamshub/flink-sql-examples/pull/55/commits/b71320b97c335f498a0409f31edfc78c6f54efb8

    user_id STRING, 
    product_id STRING, 
    quantity STRING, 
    unit_cost STRING, 
    `purchase_time` TIMESTAMP(3) METADATA FROM 'timestamp', 
    WATERMARK FOR purchase_time AS purchase_time - INTERVAL '1' SECOND 
) WITH ( 
    'connector' = 'kafka',
    'topic' = 'flink.international.sales.records', 
    'properties.bootstrap.servers' = 'my-cluster-kafka-bootstrap.flink.svc:9092', 
    'properties.group.id' = 'international-sales-record-group', 
    'value.format' = 'avro-confluent', 
    'value.avro-confluent.url' = 'http://apicurio-registry-service.flink.svc:8080/apis/ccompat/v6', 
    'scan.startup.mode' = 'latest-offset'
);
```

We can do a simple query to verify that the table was created correctly and that the data is flowing (give it a few seconds to start receiving data):

```sql
SELECT * FROM InternationalSalesRecordTable;
```

If that worked, we can now register our UDF as a [temporary catalog function](https://nightlies.apache.org/flink/flink-docs-release-2.0/docs/dev/table/sql/create/#create-function):

```sql
CREATE TEMPORARY FUNCTION currency_convert
AS 'com.github.example.CurrencyConverter'
USING JAR '/opt/currency-converter-1.0-SNAPSHOT.jar';
```

> > Note: Temporary catalog functions [only live as long as the current session](https://nightlies.apache.org/flink/flink-docs-release-2.0/docs/dev/table/functions/overview/#types-of-functions). You can omit the `TEMPORARY` keyword to create a catalog function that persists across sessions.

Reviewer: So this is absolutely correct. However, a catalog is not available by default, so we just need to be a little more explicit. Something like:
```suggestion
> Note: Temporary catalog functions [only live as long as the current session](https://nightlies.apache.org/flink/flink-docs-release-2.0/docs/dev/table/functions/overview/#types-of-functions). Provided you have a [Flink catalog](https://nightlies.apache.org/flink/flink-docs-release-2.0/docs/dev/table/catalogs/#catalogs) deployed and configured, you can omit the `TEMPORARY` keyword to create a function that persists across sessions.
```

    user_id STRING, 
    product_id STRING, 
    quantity INT, 
    iso_unit_cost STRING, 
    purchase_time TIMESTAMP(3),
    PRIMARY KEY (`user_id`) NOT ENFORCED
) WITH ( 
    'connector' = 'upsert-kafka', 
    'topic' = 'flink.iso.international.sales.records.interactive', 
    'properties.bootstrap.servers' = 'my-cluster-kafka-bootstrap.flink.svc:9092', 
    'properties.client.id' = 'sql-cleaning-client', 
    'properties.transaction.timeout.ms' = '800000', 
    'key.format' = 'csv', 
    'value.format' = 'csv', 
    'value.fields-include' = 'ALL' 
);
```

Next, let's insert the results of the formatting query into it:

```sql
INSERT INTO IsoInternationalSalesRecordTable
SELECT 
    invoice_id, 
    user_id,
    product_id,
    CAST(quantity AS INT), 
    currency_convert(unit_cost), 
    purchase_time
FROM InternationalSalesRecordTable;
```

Finally, we can verify the data is being written to the new topic by running the following command in a new terminal:

```shell
$ kubectl exec -it my-cluster-dual-role-0 -n flink -- /bin/bash \
  ./bin/kafka-console-consumer.sh --bootstrap-server localhost:9092 \
    --topic flink.iso.international.sales.records.interactive

> 5688844959819606179,user-96,0,1,"448 INR","2025-06-13 11:28:29.722"

Reviewer: It would be good to split the `<value> <currency` string into separate columns. However to do that I think you would need to create a [TableFunction](https://nightlies.apache.org/flink/flink-docs-master/docs/dev/table/functions/udfs/#table-functions) implementation and join the output with the original data. 

Perhaps this could be a future enhancement to the tutorial, it would allow you do queries on the values and currencies and do currency conversion much more easily.

### Implementing the logic

By speaking to authors of the upstream services, we should be able to obtain a list of currency symbols that can potentially appear in the `unit_cost` field:

> Here is list of currency symbols that can potentially appear in the `unit_cost` field:
>
> `'€', '₹', '₺', '฿', '₴', '₮'`
>
> — authors of the upstream services

We can create an `enum` that maps these symbols to their corresponding [ISO 4217](https://www.iso.org/iso-4217-currency-codes.html) currency codes.

As a reminder, we want to convert a string like "€100" into "100 EUR". To do this, we can use the following steps:

1. Get the first character of the string, which is the currency symbol (e.g. "€").
2. Look up the currency symbol in our `enum` to get the corresponding currency code (e.g. "€" => "EUR").
3. If the lookup failed (e.g. currency symbol was not found), we can return "ERR" as the currency code.
4. Get the rest of the string, which is the amount (e.g. "100").
5. Concatenate the currency code to the amount, and return the result (e.g. "100 EUR").

A possible implementation could look like this:

```java

> package com.github.streamshub;

Author: https://github.com/streamshub/flink-sql-examples/pull/55/commits/b76ef8b550ef098500d4571e0d505d8a1a109aad

    Currency(String symbol, String isoCode) {
        this.symbol = symbol;
        this.isoCode = isoCode;
    }

    public String getSymbol() {
        return symbol;
    }

    public String getIsoCode() {
        return isoCode;
    }

    public static Currency fromUnicodeAmount(String unicodeAmount) {
        String currencySymbol = unicodeAmount.substring(0, 1); // "€100" -> "€"
        try {
            return SYMBOL_TO_CURRENCY.getOrDefault(currencySymbol, ERR); // "€100" -> EUR
        } catch (Exception e) {
            return ERR; // "]100" -> ERR
        }
    }

    public String concatIsoCodeToAmount(String amount) {
        return amount + SEPARATOR + isoCode; // "100" + EUR -> "100 EUR"
    }

    public static String unicodeAmountToIsoAmount(String unicodeAmount) {
        String trimmedUnicodeAmount = unicodeAmount.trim();

        Currency currency = fromUnicodeAmount(trimmedUnicodeAmount); // "€100" -> EUR
        String amount = trimmedUnicodeAmount.substring(1); // "€100" -> "100"

        return currency.concatIsoCodeToAmount(amount); // "100" + EUR -> "100 EUR"
    }
}
```

We can then use this `enum` in the `eval` method of our UDF:

```java

> package com.github.streamshub;

Author: https://github.com/streamshub/flink-sql-examples/pull/55/commits/b76ef8b550ef098500d4571e0d505d8a1a109aad

> Note: This statement may succeed even if the JAR was not found or has insufficient permissions. You will likely only find this out when you try to use the UDF in a query.

Now, we can use our UDF in a query:

```sql
SELECT
    invoice_id,
    currency_convert(unit_cost) AS iso_unit_cost,
    unit_cost,
    quantity
FROM InternationalSalesRecordTable;
```

You should start seeing results with both a `unit_cost` field and an `iso_unit_cost` field containing the output of our UDF!

We can also use the UDF in more complex queries e.g. to filter for records with a specific currency and quantity:

```sql
SELECT 
    DISTINCT(product_id), 
    iso_unit_cost, 
    quantity 
  FROM (
      SELECT 
          invoice_id, 
          user_id,
          product_id,
          CAST(quantity AS INT) AS quantity, 
          currency_convert(unit_cost) AS iso_unit_cost,
          unit_cost,
          purchase_time
      FROM InternationalSalesRecordTable
    )
WHERE 
    RIGHT(iso_unit_cost, 3) = 'EUR' AND quantity > 1;
```

> > Note: This query might take a while to return results, since there are many currencies used in the data!

Reviewer: We might need to up the source generators rate as it took 30 or so seconds for this to produce a result.

</plugin>
```

We can then build the image like this:

```shell
mvn clean package docker:build
```

> Note: docker-maven-plugin uses Docker by default, if you're using Podman [you will likely need to set `DOCKER_HOST` to use podman](https://github.com/fabric8io/docker-maven-plugin/issues/1330#issuecomment-872905283).

Finally, we can create a new container using the image we just built:

```shell
# We don't need to mount the JAR anymore!
podman run -it --rm --net=host
    currency-converter:latest
        /opt/flink/bin/sql-client.sh embedded
```

You can run the same Flink SQL queries as before to verify that everything works the same way.

### Using the new UDF image in a `FlinkDeployment`

So far, we've been using the UDF in ETL queries that would have to compete for resources with other queries running in the same Flink session cluster.

Instead, like in the [Interactive ETL example](../interactive-etl/index.md), we can create a FlinkDeployment CR for deploying our queries as a stand-alone Flink Job:

```yaml
apiVersion: flink.apache.org/v1beta1
kind: FlinkDeployment
metadata:
  name: standalone-etl-udf
spec:
  # Change the two lines below depending on your image

>   image: docker.io/library/flink-sql-runner-with-currency-converter:latest

Reviewer: ```suggestion
  image: docker.io/library/flink-sql-runner-with-flink-udf-currency-converter:latest
```

  serviceAccount: flink
  jobManager:
    resource:
      memory: "2048m"
      cpu: 1
  taskManager:
    resource:
      memory: "2048m"
      cpu: 1
  job:
    jarURI: local:///opt/streamshub/flink-sql-runner.jar
    args: ["
        CREATE TABLE InternationalSalesRecordTable ( 
            invoice_id STRING, 
            user_id STRING, 
            product_id STRING, 
            quantity STRING, 
            unit_cost STRING, 
            `purchase_time` TIMESTAMP(3) METADATA FROM 'timestamp', 
            WATERMARK FOR purchase_time AS purchase_time - INTERVAL '1' SECOND 
        ) WITH ( 
            'connector' = 'kafka',
            'topic' = 'flink.international.sales.records', 
            'properties.bootstrap.servers' = 'my-cluster-kafka-bootstrap.flink.svc:9092', 
            'properties.group.id' = 'international-sales-record-group', 
            'value.format' = 'avro-confluent', 
            'value.avro-confluent.url' = 'http://apicurio-registry-service.flink.svc:8080/apis/ccompat/v6', 
            'scan.startup.mode' = 'latest-offset'
        );
        CREATE FUNCTION currency_convert
        AS 'com.github.streamshub.flink.functions.CurrencyConverter'

>         USING JAR '/opt/currency-converter-1.0-SNAPSHOT.jar';

Reviewer: ```suggestion
        USING JAR '/opt/flink-udf-currency-converter-1.0-SNAPSHOT.jar';
```

        CREATE TABLE IsoInternationalSalesRecordTable ( 
            invoice_id STRING, 
            user_id STRING, 
            product_id STRING, 
            quantity INT, 
            iso_unit_cost STRING, 
            purchase_time TIMESTAMP(3),
            PRIMARY KEY (`user_id`) NOT ENFORCED
        ) WITH ( 
            'connector' = 'upsert-kafka', 
            'topic' = 'flink.iso.international.sales.records', 
            'properties.bootstrap.servers' = 'my-cluster-kafka-bootstrap.flink.svc:9092', 
            'properties.client.id' = 'sql-cleaning-client', 
            'properties.transaction.timeout.ms' = '800000', 
            'key.format' = 'csv', 
            'value.format' = 'csv', 
            'value.fields-include' = 'ALL' 
        );
        INSERT INTO IsoInternationalSalesRecordTable
        SELECT 
            invoice_id, 
            user_id,
            product_id,
            CAST(quantity AS INT), 
            currency_convert(unit_cost), 
            purchase_time
        FROM InternationalSalesRecordTable;
        "]
    parallelism: 1
    upgradeMode: stateless
```

Then use it:

```shell
# If using minikube and a local image, load the image first:

> minikube image load flink-sql-runner-with-currency-converter

Reviewer: ```suggestion
minikube image load flink-sql-runner-with-flink-udf-currency-converter
```

    'properties.client.id' = 'sql-cleaning-client', 
    'properties.transaction.timeout.ms' = '800000', 
    'key.format' = 'csv', 
    'value.format' = 'csv', 
    'value.fields-include' = 'ALL' 
);
```

Next, let's insert the results of the formatting query into it:

```sql
INSERT INTO IsoInternationalSalesRecordTable
SELECT 
    invoice_id, 
    user_id,
    product_id,
    CAST(quantity AS INT), 
    currency_convert(unit_cost), 
    purchase_time
FROM InternationalSalesRecordTable;
```

Finally, we can verify the data is being written to the new topic by running the following command in a new terminal:

```shell
$ kubectl exec -it my-cluster-dual-role-0 -n flink -- /bin/bash \
  ./bin/kafka-console-consumer.sh --bootstrap-server localhost:9092 \
    --topic flink.iso.international.sales.records.interactive

5688844959819606179,user-96,0,1,"448 INR","2025-06-13 11:28:29.722"
7208742491425008088,user-87,106,3,"587 UAH","2025-06-13 11:28:32.725"
8796404564173987612,user-70,105,1,"399 EUR","2025-06-13 11:28:35.728"
```

## Layering the UDF on top of the `flink-sql-runner` image

### Adding [docker-maven-plugin](https://github.com/fabric8io/docker-maven-plugin)

> To make it easier for others to use our UDF, we can create a new container image that layers our JAR on top of the `flink-sql-runner` image. This way, mounting our local build of the JAR into the container will no longer be necessary.

Reviewer: ```suggestion
To make it easier for others to use our UDF, we can create a new container image that layers our JAR on top of the Streamhub [`flink-sql-runner`](https://github.com/streamshub/flink-sql) image. 
This way, the function will be available to all users of the image without needing to have the jar file locally.
It also allows us to deploy a standalone Flink job, which we will discuss later.
```

Finally, we can create a new container using the image we just built:

```shell
# We don't need to mount the JAR anymore!
podman run -it --rm --net=host \
    flink-sql-runner-with-flink-udf-currency-converter:latest \
        /opt/flink/bin/sql-client.sh embedded
```

You can run the same Flink SQL queries as before to verify that everything works the same way.

### Using the new UDF image in a `FlinkDeployment`

So far, we've been using the UDF in ETL queries that would have to compete for resources with other queries running in the same Flink session cluster.

Instead, like in the [Interactive ETL example](../interactive-etl/index.md), we can create a FlinkDeployment CR for deploying our queries as a stand-alone Flink Job:

```yaml
apiVersion: flink.apache.org/v1beta1
kind: FlinkDeployment
metadata:
  name: standalone-etl-udf
spec:
  # Change the two lines below depending on your image
  image: docker.io/library/flink-sql-runner-with-flink-udf-currency-converter:latest

>   imagePullPolicy: Never

Reviewer: ```suggestion
  # This is set to Never when you have pushed/built the image directly in your Kubernetes cluster
  imagePullPolicy: Never
```

            quantity INT, 
            iso_unit_cost STRING, 
            purchase_time TIMESTAMP(3),
            PRIMARY KEY (`user_id`) NOT ENFORCED
        ) WITH ( 
            'connector' = 'upsert-kafka', 
            'topic' = 'flink.iso.international.sales.records', 
            'properties.bootstrap.servers' = 'my-cluster-kafka-bootstrap.flink.svc:9092', 
            'properties.client.id' = 'sql-cleaning-client', 
            'properties.transaction.timeout.ms' = '800000', 
            'key.format' = 'csv', 
            'value.format' = 'csv', 
            'value.fields-include' = 'ALL' 
        );
        INSERT INTO IsoInternationalSalesRecordTable
        SELECT 
            invoice_id, 
            user_id,
            product_id,
            CAST(quantity AS INT), 
            currency_convert(unit_cost), 
            purchase_time
        FROM InternationalSalesRecordTable;
        "]
    parallelism: 1
    upgradeMode: stateless
```

Then use it:

```shell
# If using minikube and a local image, load the image first:
minikube image load flink-sql-runner-with-flink-udf-currency-converter

kubectl apply -n flink -f <path-to-flink-deployment>.yaml
```

> > Note: We can also just use the provided example FlinkDeployment CR instead:

Reviewer: ```suggestion
> Note: We can also just use the provided example FlinkDeployment CR instead (which uses the image built from the example code in this repository):
```
